diff -dur s3fs-1.61.orig/src/s3fs.cpp s3fs-1.61/src/s3fs.cpp
--- s3fs-1.61.orig/src/s3fs.cpp	2013-01-03 10:05:41.000000000 +0100
+++ s3fs-1.61/src/s3fs.cpp	2013-01-03 10:13:01.000000000 +0100
@@ -532,8 +535,8 @@
 
       // if the local and remote mtime/size
       // do not match we have an invalid cache entry
-      if(str(st.st_size) != responseHeaders["Content-Length"] || 
-        (str(st.st_mtime) != responseHeaders["x-amz-meta-mtime"])) {
+      if(st.st_size != strtoul(responseHeaders["Content-Length"].c_str(), (char **) NULL, 10) || 
+         st.st_mtime != get_mtime(responseHeaders["x-amz-meta-mtime"].c_str())) {
         if(close(fd) == -1)
           YIKES(-errno);
 
